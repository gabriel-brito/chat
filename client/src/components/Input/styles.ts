import styled, { css } from 'styled-components'
import media from 'styled-media-query'
import { InputProps } from '.'

const wrapperModifier = {
  nickname: () => css`
    border-radius: 0.3em 0 0 0.3em;
    border-right: none;
    border-left: 0.1em solid black;
    color: white;
    cursor: pointer;
    margin-right: -0.1em;
    width: 6em;

    ${media.greaterThan('medium')`
      width: 6em;
    `}

    &::placeholder {
      color: black;
      font-weight: 600;
      opacity: 1;
      text-align: center;
    }
  `
}

export const Wrapper = styled.input<Omit<InputProps, 'keyBoardFunc'>>`
  ${({ iSNicknameInput }) => css`
    border-radius: 0 0.3em 0.3em 0;
    border: 0.1em solid black;
    border-left: none;
    outline: none;
    padding: 0.5em;

    ${media.greaterThan('medium')`
      width: 85%;
    `}

    ${iSNicknameInput && wrapperModifier['nickname']}
  `}
`
