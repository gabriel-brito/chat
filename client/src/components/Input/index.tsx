import * as S from './styles'
import { ChangeEventHandler, KeyboardEventHandler } from 'react'

export type InputProps = {
  iSNicknameInput?: boolean
  placeholder?: string
  keyBoardFunc?: KeyboardEventHandler<HTMLInputElement>
  changeValueFunc?: ChangeEventHandler<HTMLInputElement>
}

const Input = ({
  changeValueFunc,
  iSNicknameInput = false,
  keyBoardFunc,
  placeholder = 'Type here to chat'
}: InputProps) => (
  <S.Wrapper
    aria-label="Input"
    autoFocus={!iSNicknameInput}
    iSNicknameInput={iSNicknameInput}
    onKeyPress={keyBoardFunc}
    onChange={changeValueFunc}
    placeholder={placeholder}
  />
)

export default Input
