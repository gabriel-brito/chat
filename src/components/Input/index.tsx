import * as S from './styles'
import { ChangeEventHandler, KeyboardEventHandler } from 'react'

export type InputProps = {
  iSNicknameInput?: boolean
  placeholder?: string
  keyBoardFunc?: KeyboardEventHandler<HTMLInputElement>
  changeValueFunc?: ChangeEventHandler<HTMLInputElement>
  value?: string
}

const Input = ({
  changeValueFunc,
  iSNicknameInput = false,
  keyBoardFunc,
  placeholder = 'Type here to chat',
  value
}: InputProps) => (
  <S.Wrapper
    aria-label="Input"
    value={value}
    autoFocus={!iSNicknameInput}
    iSNicknameInput={iSNicknameInput}
    onKeyPress={keyBoardFunc}
    onChange={changeValueFunc}
    placeholder={placeholder}
  />
)

export default Input
