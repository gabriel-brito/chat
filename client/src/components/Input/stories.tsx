import { Story, Meta } from '@storybook/react/types-6-0'
import Input, { InputProps } from '.'
import { action } from '@storybook/addon-actions'

export default {
  title: 'Input',
  component: Input
} as Meta

export const Default: Story<InputProps> = (args) => (
  <Input {...args} changeValueFunc={action('type function')} />
)

export const Nickname: Story<InputProps> = () => (
  <Input
    keyBoardFunc={action('change nick name')}
    iSNicknameInput
    placeholder="Guest"
  />
)
