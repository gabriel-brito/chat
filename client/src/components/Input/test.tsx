import { fireEvent, render, screen } from '@testing-library/react'

import Input from '.'

describe('<Input />', () => {
  const handleFunc = jest.fn()

  it('should render the input by default and match snapshot', () => {
    const { container } = render(<Input />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should call the function when user types', () => {
    render(<Input changeValueFunc={handleFunc} />)
    const input = screen.getByLabelText('Input')

    fireEvent.change(input, { target: { value: 'Hello there' } })

    expect(handleFunc).toHaveBeenCalledTimes(1)

    fireEvent.change(input, { target: { value: 'How are you going?' } })

    expect(handleFunc).toHaveBeenCalledTimes(2)
  })

  it('should render the nickname input and match snapshot', () => {
    const { container } = render(<Input iSNicknameInput />)

    expect(screen.getByLabelText('Input')).toHaveStyle({
      backgroundColor: '#283350'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should call the function when user press Enter', () => {
    render(<Input keyBoardFunc={handleFunc} />)
    const input = screen.getByLabelText('Input')

    fireEvent.keyPress(input, { key: 'Enter' })

    expect(handleFunc).toHaveBeenCalled()
  })
})
