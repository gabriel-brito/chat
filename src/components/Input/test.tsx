import { fireEvent, render, screen } from '@testing-library/react'

import Input from '.'

describe('<Input />', () => {
  const handleFunc = jest.fn()

  it('should render the input by default and match snapshot', () => {
    const { container } = render(<Input />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should call the function when user types', () => {
    render(<Input changeValueFunc={handleFunc} />)
    const input = screen.getByLabelText('Input')

    fireEvent.change(input, { target: { value: 'Hello there' } })

    expect(handleFunc).toHaveBeenCalledTimes(1)

    fireEvent.change(input, { target: { value: 'How are you going?' } })

    expect(handleFunc).toHaveBeenCalledTimes(2)
  })

  it('should render the nickname input and match snapshot', () => {
    const { container } = render(<Input iSNicknameInput />)

    expect(screen.getByLabelText('Input')).toHaveStyle({
      width: '6em'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should call the function when user press Enter', () => {
    render(<Input keyBoardFunc={handleFunc} />)
    const input = screen.getByLabelText('Input')

    fireEvent.keyPress(input, { key: 'Enter' })

    expect(handleFunc).toHaveBeenCalled()
  })
})
