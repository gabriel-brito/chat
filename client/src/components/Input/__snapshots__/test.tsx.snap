// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render the input by default and match snapshot 1`] = `
.c0 {
  border-radius: 0 0.3em 0.3em 0;
  border: 0.1em solid black;
  border-left: none;
  outline: none;
  padding: 0.5em;
}

@media (min-width:768px) {
  .c0 {
    width: 85%;
  }
}

<input
  aria-label="Input"
  class="c0"
  placeholder="Type here to chat"
/>
`;

exports[`<Input /> should render the nickname input and match snapshot 1`] = `
.c0 {
  border-radius: 0 0.3em 0.3em 0;
  border: 0.1em solid black;
  border-left: none;
  outline: none;
  padding: 0.5em;
  border-radius: 0.3em 0 0 0.3em;
  border-right: none;
  border-left: 0.1em solid black;
  color: white;
  cursor: pointer;
  margin-right: -0.1em;
  width: 6em;
}

.c0::-webkit-input-placeholder {
  color: black;
  font-weight: 600;
  opacity: 1;
  text-align: center;
}

.c0::-moz-placeholder {
  color: black;
  font-weight: 600;
  opacity: 1;
  text-align: center;
}

.c0:-ms-input-placeholder {
  color: black;
  font-weight: 600;
  opacity: 1;
  text-align: center;
}

.c0::placeholder {
  color: black;
  font-weight: 600;
  opacity: 1;
  text-align: center;
}

@media (min-width:768px) {
  .c0 {
    width: 85%;
  }
}

@media (min-width:768px) {
  .c0 {
    width: 6em;
  }
}

<input
  aria-label="Input"
  class="c0"
  placeholder="Type here to chat"
/>
`;
